src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/ScalaBuildServerService.java

```diff
@@ -39,19 +54,65 @@ public ScalaBuildServerService(BazelData bazelData, BazelRunner bazelRunner) {
       ScalacOptionsParams scalacOptionsParams) {
-
     List<ScalacOptionsItem> resultItems =
         targetsLanguageOptionsResolver.getResultItemsForTargets(scalacOptionsParams.getTargets());
 
     ScalacOptionsResult javacOptionsResult = new ScalacOptionsResult(resultItems);
     return Either.forRight(javacOptionsResult);
   }
 
-  public CompletableFuture<ScalaTestClassesResult> buildTargetScalaTestClasses(
+  public Either<ResponseError, ScalaTestClassesResult> buildTargetScalaTestClasses(
       ScalaTestClassesParams scalaTestClassesParams) {
-    System.out.printf("DWH: Got buildTargetScalaTestClasses: %s%n", scalaTestClassesParams);
-    // TODO(illicitonion): Populate
-    return CompletableFuture.completedFuture(new ScalaTestClassesResult(new ArrayList<>()));
+    TargetRulesResolver<ScalaTestClassesItem> targetRulesResolver =
+        TargetRulesResolver.withBazelRunnerAndFilterAndMapper(
+            bazelRunner, this::doesAttributesContainTestClass, this::mapRuleToTestClassesItem);
+
+    List<ScalaTestClassesItem> resultItems =
+        targetRulesResolver.getItemsForTargets(scalaTestClassesParams.getTargets());
+
+    ScalaTestClassesResult scalaTestClassesResult = new ScalaTestClassesResult(resultItems);
+
+    return Either.forRight(scalaTestClassesResult);
+  }
+
+  private ScalaTestClassesItem mapRuleToTestClassesItem(Build.Rule rule) {
+    BuildTargetIdentifier target = new BuildTargetIdentifier(rule.getName());
+    List<String> classes = getTestClasses(rule);
+
+    return new ScalaTestClassesItem(target, classes);
+  }
+
+  private List<String> getTestClasses(Build.Rule rule) {
+    List<String> mainClasses = getTestMainClasses(rule);
+    List<String> srscClasses = getTestSrcsClasses(rule);
+
+    return Stream.concat(mainClasses.stream(), srscClasses.stream()).collect(Collectors.toList());
+  }
+
+  private List<String> getTestMainClasses(Build.Rule rule) {
+    return rule.getAttributeList().stream()
+        .filter(
+            attribute ->
+                TargetsUtils.isAttributeSpecifiedAndHasGivenName(
+                    attribute, SCALA_TEST_MAIN_CLASSES_ATTRIBUTE_NAME))
+        .map(Attribute::getStringValue)
+        .collect(Collectors.toList());
+  }
+
+  private List<String> getTestSrcsClasses(Build.Rule rule) {
+    return rule.getAttributeList().stream()
+        .filter(
+            attribute ->
+                TargetsUtils.isAttributeSpecifiedAndHasGivenName(
+                    attribute, SCALA_TEST_SRCS_CLASSES_ATTRIBUTE_NAME))
+        .map(Attribute::getStringListValueList)
+        .flatMap(Collection::stream)
+        .collect(Collectors.toList());
+  }
+
+  private boolean doesAttributesContainTestClass(Build.Rule rule) {
+    return TargetsUtils.doesRuleAttributesContain(rule, SCALA_TEST_MAIN_CLASSES_ATTRIBUTE_NAME)
+        || TargetsUtils.doesRuleAttributesContain(rule, SCALA_TEST_SRCS_CLASSES_ATTRIBUTE_NAME);
   }
```

Comment:
Shouldn't we check name of the rule in order to find test classes? ('scala_test' probably?)

Reply:
right! I forgot to check it